contribution-guidelines.md

```diff
@@ -0,0 +1,147 @@
+the case that we can report the error to the caller, but some cleanup has to be done beforehand.
+
+Sometimes, the service itself cannot respond to an error condition. For example, if a host that a service is running on
+encounters a network failure. In these cases, the service should be restarted automatically as soon as possible. For
+cases where the host itself encounters an error, merely having the service running inside Kubernetes will help. For
+other failure conditions, we have to provide a way for Kubernetes to detect when something is wrong with the
+service. Having useful and reliable health and readiness checks for each service will help to ensure that services will
+be restarted if something goes wrong.
+
+## Debugability
+
+Debugability means that callers should be able to identify and respond to errors. For REST interfaces, this means that
+HTTP status codes should be correct and that error response bodies should contain enough information for callers to
+identify and correct the problem. For example, if a resource isn't found, an HTTP endpoint should return status
+code 404. If there's a problem with the request then the status code should be in the 400 series, and a response body
+providing additional information about the error should be returned. For new endpoints, the response body should look
+like this:
+
+``` JavaScript
+{
+    "message": "error details here"
+}
+```
+
+Additional fields may be present, but the `message` field should always be included. Note that this is the default error
+response body format used by [labstack/echo][10].
+
+## Performance
+
+The Discovery Environemnt sometimes makes multiple service calls to perform a single task, so it's important that every
+endpoint in Terrain is as responsive as reasonably possible. We say "reasonably possible" because small performance
+improvements can sometimes come at too high of a cost in terms of development effort. As a general rule of thumb,
+endpoints should return within 2 seconds whenever possible. When considering performance (or looking for ways to improve
+the performance of an endoint), consider the following questions:
+
+- Is the endpoint duplicating work? For example, is the same value calculated in multiple different places? Can the
+  endpoint benefit from a caching strategy of some sort?
+- Are individual calculations being performed when a bulk calculation might be faster? For example, are multiple
+  database calls being made when a single call could be used more efficiently?
+- Can some tasks be done in parallel? A classic example of this is when Terrain is orchestrating calls to other
+  services, and the API calls are not dependent on each other.
+- Is a service endpoint doing too much? Sometimes its better to split up an endpoint so that clients can adjust how the
+  tasks are being performed. For example, maybe an endpoint is always returning some piece of information that's only
+  used in a small number of cases.
```

Comment:
Do we want to mention any tools to help with analyzing/debugging performance?  Not sure if we're pretty sold on opentelemetry for example.

Reply:
That's a good idea as well. I admittedly don't have much experience with opentelemetry, but I know that @ianmcorvidae has found it to be extremely useful. It might be useful to add a blurb mentioning it here.

Reply:
I haven't written much... anything about opentelemetry, but I did just open https://github.com/cyverse-de/terrain/pull/194 with rudimentary opentelemetry stuff in this repository. We talked about me doing a bit of a talk about opentelemetry at some point, but I can also help if it'd be useful in adding any information about it here.

Reply:
That would definitely be helpful. I'll merge this soon so that you can rebase your changes and update this document. Thanks!